### ISSUE TITLE
1.2.1 build failure on Heroku

### ISSUE THREAD

**Issue by erhhung** [Automattic/node-canvas#526]
First of all, release 1.1.6 builds just fine. I've been building my app on that version for a while now. I use `https://github.com/mojodna/heroku-buildpack-cairo` to install Cairo, with no additional buildpacks or packages.

I then upgraded to 1.2.1, and now I'm getting this build output (I'm using the Heroku Cedar stack, BTW, but I also tried on Cedar-14 and got the same failures):

````
       > canvas@1.2.1 install /tmp/build_9bfd70b688d40f7cacdcf6c6ad953ca2/node_modules/canvas
       > node-gyp rebuild
       
       Package pangocairo was not found in the pkg-config search path.
       Perhaps you should add the directory containing `pangocairo.pc'
       to the PKG_CONFIG_PATH environment variable
       No package 'pangocairo' found
       gyp: Call to 'pkg-config pangocairo --libs' returned exit status 1. while trying to load binding.gyp
       gyp ERR! configure error 
       gyp ERR! stack Error: `gyp` failed with exit code: 1
       gyp ERR! stack     at ChildProcess.onCpExit (/tmp/build_9bfd70b688d40f7cacdcf6c6ad953ca2/.heroku/node/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:343:16)
       gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
       gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
       gyp ERR! System Linux 3.13.0-40-generic
       gyp ERR! command "node" "/tmp/build_9bfd70b688d40f7cacdcf6c6ad953ca2/.heroku/node/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
       gyp ERR! cwd /tmp/build_9bfd70b688d40f7cacdcf6c6ad953ca2/node_modules/canvas
       gyp ERR! node -v v0.10.36
       gyp ERR! node-gyp -v v1.0.1
       gyp ERR! not ok 
````

I checked the build environment, and `pangocairo.pc` was no where to be found, so I tried adding pango using several methods to no avail. I first tried installing `https://github.com/ddollar/heroku-buildpack-apt` with `Aptfile` containing `build-essential g++ libfontconfig1-dev libpango1.0-dev libcairo2-dev libjpeg8-dev libgif-dev`, but got this output:

````
       > canvas@1.2.1 install /tmp/build_05710aa88015453db8bf0f02acacd84e/node_modules/canvas
       > node-gyp rebuild
       
       make: Entering directory `/tmp/build_05710aa88015453db8bf0f02acacd84e/node_modules/canvas/build'
         SOLINK_MODULE(target) Release/obj.target/canvas-postbuild.node
         SOLINK_MODULE(target) Release/obj.target/canvas-postbuild.node: Finished
         COPY Release/canvas-postbuild.node
         CXX(target) Release/obj.target/canvas/src/Canvas.o
       In file included from ../src/Canvas.cc:7:
       ../src/Canvas.h:17:30: warning: pango/pangocairo.h: No such file or directory
       In file included from ../src/Canvas.cc:7:
       ../src/Canvas.h:66: error: ‘cairo_status_t’ has not been declared
       ../src/Canvas.h:83: error: ISO C++ forbids declaration of ‘cairo_surface_t’ with no type
       ../src/Canvas.h:83: error: ‘cairo_surface_t’ declared as an ‘inline’ field
       ../src/Canvas.h:83: error: expected ‘;’ before ‘*’ token
       ../src/Canvas.h:84: error: expected ‘;’ before ‘inline’
       ../src/Canvas.h:92: error: ISO C++ forbids declaration of ‘cairo_surface_t’ with no type
       ../src/Canvas.h:92: error: expected ‘;’ before ‘*’ token
       ../src/Canvas.h: In member function ‘uint8_t* Canvas::data()’:
       ../src/Canvas.h:85: error: ‘_surface’ was not declared in this scope
       ../src/Canvas.h:85: error: ‘cairo_image_surface_get_data’ was not declared in this scope
       ../src/Canvas.h: In member function ‘int Canvas::stride()’:
       ../src/Canvas.h:86: error: ‘_surface’ was not declared in this scope
       ../src/Canvas.h:86: error: ‘cairo_image_surface_get_stride’ was not declared in this scope
       In file included from ../src/Canvas.cc:8:
       ../src/PNG.h: In function ‘cairo_status_t canvas_write_png(cairo_surface_t*, void (*)(png_struct*, png_byte*, png_size_t), void*)’:
       ../src/PNG.h:165: warning: case label value is less than minimum value for type
       ../src/PNG.h:166: warning: case label value exceeds maximum value for type
       In file included from ../src/Canvas.cc:9:
       ../src/CanvasRenderingContext2d.h: At global scope:
       ../src/CanvasRenderingContext2d.h:54: error: ‘PangoWeight’ does not name a type
       ../src/CanvasRenderingContext2d.h:55: error: ‘PangoStyle’ does not name a type
       ../src/CanvasRenderingContext2d.h:170: error: ISO C++ forbids declaration of ‘PangoLayout’ with no type
       ../src/CanvasRenderingContext2d.h:170: error: ‘PangoLayout’ declared as an ‘inline’ field
       ../src/CanvasRenderingContext2d.h:170: error: expected ‘;’ before ‘*’ token
       ../src/CanvasRenderingContext2d.h:173: error: expected ‘;’ before ‘private’
       ../src/CanvasRenderingContext2d.h:179: error: ISO C++ forbids declaration of ‘PangoLayout’ with no type
       ../src/CanvasRenderingContext2d.h:179: error: expected ‘;’ before ‘*’ token
       ../src/Canvas.cc: In static member function ‘static void Canvas::ToBufferAsync(uv_work_t*)’:
       ../src/Canvas.cc:182: error: ‘class Canvas’ has no member named ‘surface’
       ../src/Canvas.cc: In static member function ‘static v8::Handle<v8::Value> Canvas::ToBuffer(const v8::Arguments&)’:
       ../src/Canvas.cc:245: error: ‘class Canvas’ has no member named ‘surface’
       ../src/Canvas.cc:328: error: ‘class Canvas’ has no member named ‘surface’
       ../src/Canvas.cc: In static member function ‘static v8::Handle<v8::Value> Canvas::StreamPNGSync(const v8::Arguments&)’:
       ../src/Canvas.cc:420: error: ‘class Canvas’ has no member named ‘surface’
       ../src/Canvas.cc: In static member function ‘static v8::Handle<v8::Value> Canvas::StreamJPEGSync(const v8::Arguments&)’:
       ../src/Canvas.cc:460: error: ‘class Canvas’ has no member named ‘surface’
       ../src/Canvas.cc: In constructor ‘Canvas::Canvas(int, int, canvas_type_t)’:
       ../src/Canvas.cc:477: error: ‘_surface’ was not declared in this scope
       ../src/Canvas.cc: In destructor ‘virtual Canvas::~Canvas()’:
       ../src/Canvas.cc:507: error: ‘_surface’ was not declared in this scope
       ../src/Canvas.cc: In member function ‘void Canvas::resurface(v8::Handle<v8::Object>)’:
       ../src/Canvas.cc:529: error: ‘_surface’ was not declared in this scope
       ../src/Canvas.cc:544: error: ‘surface’ was not declared in this scope
       ../src/Canvas.cc:561: error: ‘surface’ was not declared in this scope
       ../src/Canvas.cc: At global scope:
       ../src/Canvas.cc:573: error: prototype for ‘v8::Local<v8::Value> Canvas::Error(cairo_status_t)’ does not match any in class ‘Canvas’
       ../src/Canvas.h:66: error: candidate is: static v8::Local<v8::Value> Canvas::Error(int)
       make: *** [Release/obj.target/canvas/src/Canvas.o] Error 1
       make: Leaving directory `/tmp/build_05710aa88015453db8bf0f02acacd84e/node_modules/canvas/build'
       gyp ERR! build error 
       gyp ERR! stack Error: `make` failed with exit code: 2
       gyp ERR! stack     at ChildProcess.onExit (/tmp/build_05710aa88015453db8bf0f02acacd84e/.heroku/node/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
       gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
       gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
       gyp ERR! System Linux 3.13.0-40-generic
       gyp ERR! command "node" "/tmp/build_05710aa88015453db8bf0f02acacd84e/.heroku/node/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
       gyp ERR! cwd /tmp/build_05710aa88015453db8bf0f02acacd84e/node_modules/canvas
       gyp ERR! node -v v0.10.36
       gyp ERR! node-gyp -v v1.0.1
       gyp ERR! not ok 
````

I thought that all the dev packages would have `pangocairo.h`, but maybe not! So then I found and installed `https://github.com/bkeepers/heroku-buildpack-pango` in addition to `https://github.com/mojodna/heroku-buildpack-cairo`. That produced this output:

````
       > canvas@1.2.1 install /tmp/build_57f58adf936b9193d4f6edde815f72d0/node_modules/canvas
       > node-gyp rebuild
       
       make: Entering directory `/tmp/build_57f58adf936b9193d4f6edde815f72d0/node_modules/canvas/build'
         SOLINK_MODULE(target) Release/obj.target/canvas-postbuild.node
         SOLINK_MODULE(target) Release/obj.target/canvas-postbuild.node: Finished
         COPY Release/canvas-postbuild.node
         CXX(target) Release/obj.target/canvas/src/Canvas.o
         CXX(target) Release/obj.target/canvas/src/CanvasGradient.o
         CXX(target) Release/obj.target/canvas/src/CanvasPattern.o
         CXX(target) Release/obj.target/canvas/src/CanvasRenderingContext2d.o
       ../src/CanvasRenderingContext2d.cc: In member function ‘void Context2d::shadow(void (*)(cairo_t*))’:
       ../src/CanvasRenderingContext2d.cc:374: warning: unused variable ‘surface’
         CXX(target) Release/obj.target/canvas/src/color.o
         CXX(target) Release/obj.target/canvas/src/Image.o
         CXX(target) Release/obj.target/canvas/src/ImageData.o
         CXX(target) Release/obj.target/canvas/src/init.o
         CXX(target) Release/obj.target/canvas/src/PixelArray.o
         CXX(target) Release/obj.target/canvas/src/FontFace.o
         SOLINK_MODULE(target) Release/obj.target/canvas.node
       /usr/bin/ld: /tmp/build_57f58adf936b9193d4f6edde815f72d0/.apt/usr/lib/../lib/libcairo.a(cairo.o): relocation R_X86_64_32 against `_cairo_pattern_nil' can not be used when making a shared object; recompile with -fPIC
       /tmp/build_57f58adf936b9193d4f6edde815f72d0/.apt/usr/lib/../lib/libcairo.a: could not read symbols: Bad value
       collect2: ld returned 1 exit status
       make: *** [Release/obj.target/canvas.node] Error 1
       make: Leaving directory `/tmp/build_57f58adf936b9193d4f6edde815f72d0/node_modules/canvas/build'
       gyp ERR! build error 
       gyp ERR! stack Error: `make` failed with exit code: 2
       gyp ERR! stack     at ChildProcess.onExit (/tmp/build_57f58adf936b9193d4f6edde815f72d0/.heroku/node/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
       gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
       gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
       gyp ERR! System Linux 3.13.0-40-generic
       gyp ERR! command "node" "/tmp/build_57f58adf936b9193d4f6edde815f72d0/.heroku/node/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
       gyp ERR! cwd /tmp/build_57f58adf936b9193d4f6edde815f72d0/node_modules/canvas
       gyp ERR! node -v v0.10.36
       gyp ERR! node-gyp -v v1.0.1
       gyp ERR! not ok 
````

Not sure what else to try now, so I've reverted back to using 1.1.6.

**Comment by lpinca** [comment]
I received the same build error (the first one) on a CentOS 6 machine.
It was fixed by installing `pango-devel` but it seems that the [`has_lib`](https://github.com/Automattic/node-canvas/blob/1c1d93e6177fca306a22b81c3650a253782aadc8/binding.gyp#L16) util returns a false positive.

This is the [commit](https://github.com/Automattic/node-canvas/commit/7522aca2ea7d4e38752cfef807b31ba1a419d58f?diff=unified#diff-21e418851ae48f9b11907837add9e39a) that introduced pango support.

**Comment by MattouFP** [comment]
Same problem here, shouldn't pango be optional?

**Comment by LinusU** [comment]
I think that I recently fixed this with #701 

Closing this since it's an old thread, please see our new installation instructions and report back if you still have problems...
